aspnetcore/performance/performance-best-practices.md

```diff
@@ -0,0 +1,136 @@
+* Code that is executed for every request or multiple times per request. For example, custom logging, authorization handlers, or initialization of transient services.
+
+Recommendations:
+
+* **Do not** use custom middleware components with long-running tasks.
+* **Do** use performance profiling tools (like [Visual Studio Diagnostic Tools](/visualstudio/profiling/profiling-feature-tour) or [PerfView](https://github.com/Microsoft/perfview)) to identify [hot code paths](#hot).
+
+## Complete long-running Tasks outside of HTTP requests
+
+Most requests to an ASP.NET Core app can be handled by a controller or page model calling necessary services and returning an HTTP response. For some requests which involve long-running tasks, it's better to make the entire request-response process asynchronous.
+
+Recommendations:
+
+* **Do not** wait for long-running tasks to complete as part of ordinary HTTP request processing.
+* **Do** consider handling long-running requests with [background services](/aspnet/core/fundamentals/host/hosted-services) or out of process with an [Azure Function](/azure/azure-functions/). Completing work out-of-process is especially beneficial for CPU-intensive tasks.
+* **Do** use real-time communication options like [SignalR](xref:signalr/introduction) to communicate with clients asynchronously.
+
+## Minify client assets
+
+ASP.NET Core apps with complex front-ends frequently serve many JavaScript, CSS, or image files. Performance of initial load requests can be improved by:
+
+* Bundling, which combines multiple files into one.
+* Minifying, which reduces the size of files by.
+
+Recommendations:
+
+* **Do** use ASP.NET Core's [built-in support](xref:client-side/bundling-and-minification) for bundling and minifying client assets.
+* **Do** consider other third-party tools like [Gulp](uid:client-side/bundling-and-minification#consume-bundleconfigjson-from-gulp) or [Webpack](https://webpack.js.org/) for more complex client asset management.
+
+## Use the latest ASP.NET Core release
+
+Each new release of ASP.NET includes performance improvements. Optimizations in .NET Core and ASP.NET Core mean that newer versions will outperform older versions. For example, .NET Core 2.1 added support for compiled regular expressions and benefitted from [`Span<T>`](https://msdn.microsoft.com/magazine/mt814808.aspx). ASP.NET Core 2.2 added support for HTTP/2. If performance is a priority, consider upgrading to the most current version of ASP.NET Core.
+
+<!-- TODO review link and taking advantage of new [performance features](#TBD)
+Maybe skip this TBD link as each version will have perf improvements -->
+
+## Minimize exceptions
+
+Exceptions should be rare. Avoid using exceptions as a way to control normal program flow. If possible, include logic in the app code to detect and handle conditions that would cause an exception. For example, log problems rather than throwing an exception.
```

Comment:
This is a good addition. It doesn't follow the do/do-not pattern of many of the other recommendations, though. The 'Use the latest ASP.NET Core release' one doesn't either but it feels different somehow. It may be worthwhile moving this recommendation about the 'use the latest ASP.NET Core' one and re-formatting it to match the pattern of the others:

## Minimize exceptions

Exceptions should be rare. Throwing and catching exceptions is slow relative to other code flow patterns. Because of this, exceptions should not be used to control normal program flow.

Recommendations:

* **Do not** use throwing or catching exceptions as a means of normal program flow, especially in hot code paths.
* **Do** include logic in the app to detect and handle conditions that would cause an exception.
* **Do** throw or catch exceptions which may be related to unusual or unexpected conditions.

App diagnostic tools (like Application Insights) can help to identify common exceptions in an application which may affect performance.